orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/services/ClientSignatureValidationService.java

```diff
@@ -78,14 +119,80 @@ public void validateTokenClientAssertion(PrivateKeyJWT privateKeyJWT, ClientRegi
+        try {
+            if (client.getPublicKeySource().equals(PublicKeySource.STATIC.getValue())) {
+                return convertPemToPublicKey(client.getPublicKey());
+            }
+            if (kid == null) {
+                String error = "Key ID is null but is required to fetch JWKS";
+                LOG.error(error);
+                throw new ClientSignatureValidationException(error);
+            }
+            String jwksUrl = client.getJwksUrl();
+            if (client.getJwksUrl() == null) {
+                String error = "Client JWKS URL is null but is required to fetch JWKS";
+                LOG.error(error);
+                throw new ClientSignatureValidationException(error);
+            }
+            Optional<RpPublicKeyCache> cache =
+                    rpPublicKeyCacheService.getRpPublicKeyCacheData(client.getClientID(), kid);
+            if (cache.isPresent()) {
+                return JWK.parse(cache.get().getPublicKey()).toRSAKey().toPublicKey();
+            }
+
+            InvokeResponse response = invokeFetchJwksFunction(lambdaClient, jwksUrl, kid);
+            String unescapedPayload =
+                    objectMapper.readValue(response.payload().asUtf8String(), String.class);
+            if (unescapedPayload.equals("error")) {
+                String error = "Returned error from FetchJwksHandler";
+                LOG.error(error);
+                throw new ClientSignatureValidationException(error);
+            }
+
+            JWK jwk = JWK.parse(unescapedPayload);
+            rpPublicKeyCacheService.addRpPublicKeyCacheData(
+                    client.getClientID(), jwk.getKeyID(), jwk.toJSONString());
+            return jwk.toRSAKey().toPublicKey();
+        } catch (ParseException | Json.JsonException e) {
+            throw new RuntimeException();
+        } catch (JOSEException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static PublicKey convertPemToPublicKey(String publicKeyPem)
             throws NoSuchAlgorithmException, InvalidKeySpecException {
-        byte[] decodedKey = Base64.getMimeDecoder().decode(client.getPublicKey());
+        byte[] decodedKey = Base64.getMimeDecoder().decode(publicKeyPem);
         X509EncodedKeySpec keySpec = new X509EncodedKeySpec(decodedKey);
         KeyFactory kf = KeyFactory.getInstance(KeyType.RSA.getValue());
         return kf.generatePublic(keySpec);
     }
 
+    private InvokeResponse invokeFetchJwksFunction(
+            LambdaClient awsLambda, String jwksUrl, String kid) {
+        try {
+            JSONObject jsonObj = new JSONObject();
+            jsonObj.put("url", jwksUrl);
+            jsonObj.put("keyId", kid);
+            String json = jsonObj.toString();
+            SdkBytes payload = SdkBytes.fromUtf8String(json);
+
+            InvokeRequest request =
+                    InvokeRequest.builder()
+                            .functionName(
```

Comment:
As elsewhere, we want to be invoking the alias so we have control over release and rollback